src/compositor/compositor-xrender.c

```diff
@@ -833,10 +833,10 @@ solid_picture (MetaDisplay *display,
       return None;
     }
 
-  c.alpha = a * 0xffff;
-  c.red = r * 0xffff;
-  c.green = g * 0xffff;
-  c.blue = b * 0xffff;
+  c.alpha = (guint16)(a * 65535 + 0.5);
```

Comment:
Why `+ 0.5` everywhere?

Reply:
It's the recommended and the generic way to convert floating point numbers to integers.

Reply:
Who recommends that? Generic way?

Reply:
It recall that it's the JTC1/SC22/WG11, but I may be wrong.

Reply:
How would you do it? truncating?

Reply:
When you do a floating point conversion to an integer, you can represent it as a staircase on a graph, where each step is the integer value. You have to take into account the number of digits you have for accuracy (floating point number representation), and the range of the interval (number of steeps). If you have many step, the conversion error will be smaller or if you have a lot of accuracy (sampling theory).

Truncating means to remove the decimal part from a number. Truncating a scaling result (max_value * factor) means to remove the last step of the staircase, for this reason rounding is done (+ 0.5). 
```
integer_value = max_value * factor, 0 <= factor <= 1
```

Reply:
Anyway, I think explaining this doesn’t make much sense here, and I guess there are many others who can explain it better than me. :)

Reply:
Now that we talk about it, I don't know why glib developers always use double precision floating point number for graphics, penalizing devices with limited capabilities like on the raspberry pi platform. But I suppose we have to live with this choice ... I guess someone got it wrong and did it thinking it works just like going from a 32 to 64 bit architecture.

Reply:
I think change **without** explanation does not make any sense... Sorry, but this is first time that I hear that using `+ 0.5` is recommended way to convert float to int as generic way.

Adding `+ 0.5` everywhere is not needed to fix warning you are fixing with this MR. You are mixing warning fix with rounding behavior changes. What bug/issue are you trying to solve with this change?

In one comment you mention performance as reason and yet you make CPU do more work for no real reason.

Reply:
I have not invented anything new, color scaling function on mate-desktop/libmate-desktop/mate-colorsel.c 
```C
/* Conversion between 0->1 double and and guint16. See
 * scale_round() below for more general conversions
 */
#define SCALE(i) (i / 65535.)
#define UNSCALE(d) ((guint16)(d * 65535 + 0.5))
#define INTENSITY(r, g, b) ((r) * 0.30 + (g) * 0.59 + (b) * 0.11)
```

Reply:
That was added as bug fix - https://bugzilla.gnome.org/show_bug.cgi?id=58120. What are you fixing here?

Reply:
Yes, you are right. There is no need to make any rounding, ie it's not necessary to override the default gcc method for rounding floating point numbers which may vary depending on the FPU used by the system.

Test:
```C
/* gcc -Wfloat-conversion test2.c -o test2 */
#include <stdio.h>
#include <limits.h>

int main (int argc, char **argv)
{
  unsigned short i=0, temp;
  double d;

  do {
    d = (double)i/USHRT_MAX;
    temp = (unsigned short)(d * USHRT_MAX);
    if (i != temp)
      printf ("%hu %hu%s\n", i, temp, i-temp != 0 ? " x" : "");
    i++;
  } while (i!=0);

  return 0;
}
```

Reply:
I got the same result if simple precision is used instead of double precision which usually is more expensive.
```C
/* gcc -Wfloat-conversion test2.c -o test2 */
#include <stdio.h>
#include <limits.h>

int main (int argc, char **argv)
{
  unsigned short i=0, temp;
  float d;

  do {
    d = (float)i/USHRT_MAX;
    temp = (unsigned short)(d * USHRT_MAX);
    if (i != temp)
      printf ("%hu %hu%s\n", i, temp, i-temp != 0 ? " x" : "");
    i++;
  } while (i!=0);

  return 0;
}
```